mypy issues on `main`

The `pre-commit` is reporting some `mypy` issues on the `main` branch:
```
xarray/core/formatting.py:193: error: No overload variant of "max" matches argument types "signedinteger[Any]", "int"  [call-overload]
xarray/core/formatting.py:193: note: Possible overload variant:
xarray/core/formatting.py:193: note:     def [SupportsLessThanT <: SupportsLessThan] max(__arg1, SupportsLessThanT, SupportsLessThanT, *_args: SupportsLessThanT, key: None = ...) -> SupportsLessThanT
xarray/core/formatting.py:193: note:     <5 more non-matching overloads not shown>
xarray/core/indexes.py:166: error: Incompatible types in assignment (expression has type "Union[dtype[object_], dtype[Any], dtype[void]]", variable has type "dtype[object_]")  [assignment]
xarray/core/computation.py:607: error: Argument 1 to "append" of "list" has incompatible type "None"; expected "slice"  [arg-type]
xarray/core/accessor_str.py:277: error: <nothing> not callable  [misc]
Found 4 errors in 4 files (checked 143 source files)
```

does anyone know how to fix those?

--- a/xarray/core/accessor_str.py
+++ b/xarray/core/accessor_str.py
@@ -270,7 +270,10 @@ class StringAccessor:
 
         if getattr(pat, "dtype", None) != np.object_:
             pat = self._stringify(pat)
-        func = lambda x: re.compile(x, flags=flags)
+
+        def func(x):
+            return re.compile(x, flags=flags)
+
         if isinstance(pat, np.ndarray):
             # apply_ufunc doesn't work for numpy arrays with output object dtypes
             func = np.vectorize(func)

--- a/xarray/core/computation.py
+++ b/xarray/core/computation.py
@@ -597,7 +597,7 @@ def broadcast_compat_data(
         data = duck_array_ops.transpose(data, order)
 
     if new_dims != reordered_dims:
-        key_parts: list[Union[slice, None]] = []
+        key_parts: List[Optional[slice]] = []
         for dim in new_dims:
             if dim in set_old_dims:
                 key_parts.append(SLICE_NONE)

--- a/xarray/core/computation.py
+++ b/xarray/core/computation.py
@@ -597,7 +597,7 @@ def broadcast_compat_data(
         data = duck_array_ops.transpose(data, order)
 
     if new_dims != reordered_dims:
-        key_parts = []
+        key_parts: list[Union[slice, None]] = []
         for dim in new_dims:
             if dim in set_old_dims:
                 key_parts.append(SLICE_NONE)

--- a/xarray/core/formatting.py
+++ b/xarray/core/formatting.py
@@ -190,7 +190,7 @@ def format_array_flat(array, max_width: int):
         (max_possibly_relevant < array.size) or (cum_len > max_width).any()
     ):
         padding = " ... "
-        max_len = max(np.argmax(cum_len + len(padding) - 1 > max_width), 2)  # type: ignore[type-var]
+        max_len = max(int(np.argmax(cum_len + len(padding) - 1 > max_width)), 2)  # type: ignore[type-var]
         count = min(array.size, max_len)
     else:
         count = array.size

--- a/xarray/core/indexes.py
+++ b/xarray/core/indexes.py
@@ -163,7 +163,7 @@ class PandasIndex(Index, ExplicitlyIndexedNDArrayMixin):
             else:
                 dtype_ = array.dtype
         else:
-            dtype_ = np.dtype(dtype)
+            dtype_ = np.dtype(dtype)  # type: ignore[assignment]
         self._dtype = dtype_
 
     def to_pandas_index(self) -> pd.Index:

